### PR TITLE
build: Do not build with Remotery by default

### PR DESCRIPTION
**Pull request by jjerphan**
#### Reference Issues/PRs

Extracted from #449.

#### What does this implement/fix? Explain your changes.

Remotery seems to be used as a performance analysis tool for development.
Remotery might not be needed when it is released to users.

This PR proposes to add a `BUILD_WITH_REMOTERY` CMake option (`OFF` by default) to build ArcticDB with Remotery if needed. It reuses the existing macro.

#### Any other comments?

ArcticDB for conda-forge will need not be built with Remotery. 
Should we turn `BUILD_WITH_REMOTERY` to `ON` by default?

What do you think, @mehertz and @GemDot?

### CODE REVIEW

**Review by mehertz — APPROVED** [review]

**Comment by jjerphan** [comment]
I would prefer waiting for another reviewer to give their approval and to merge rather that merging this PR myself in order to not to miss something which does not appear in the hunks' context.

File: cpp/arcticdb/entity/performance_tracing.hpp
Comment (jjerphan) [reply cpp/arcticdb/entity/performance_tracing.hpp:122]: Yes, but they are needed in the code-base when we build without Remotery (I can add a commit to mention that).

Comment (qc00) [reply cpp/arcticdb/entity/performance_tracing.hpp:122]: No, I meant the opposite. The functionality of these two does not involve Remotery and should not be replaced with these no-op macros when Remotery is disabled.

In fact, the `ARCTICDB_SAMPLE*` macros had a non-Remotery mode as well.

File: cpp/arcticdb/entity/performance_tracing.cpp
Comment (jjerphan) [reply cpp/arcticdb/entity/performance_tracing.cpp:16]: Here are the occurrences of the class outside of this file:

```
cpp/arcticdb/python/python_module.cpp:234:    py::class_<RemoteryInstance, std::shared_ptr<RemoteryInstance>>(remotery, "Instance");
cpp/arcticdb/util/global_lifetimes.cpp:30:    RemoteryInstance::destroy_instance();
cpp/arcticdb/version/local_versioned_engine.cpp:35:    auto temp = RemoteryInstance::instance();
```

What would you recommend doing instead? To conditionally define the class to implement its methods when `USE_REMOTERY` is defined?

Comment (qc00) [reply cpp/arcticdb/entity/performance_tracing.cpp:16]: I was thinking of putting the `#if` in the body of the `RemoteryInstance` methods and the `rmt_` field, so they do nothing if Remotery is not enabled, but the class still exists.

That way this PR can be reduced to a couple of files.

Comment (jjerphan) [reply cpp/arcticdb/entity/performance_tracing.cpp:16]: I thought of this alternative but eventually did not choose it as I thought this could mislead readers.

The current solution is touching a few files but I think is a bit more explicit, and does not create unusable objects.

What do you think?

Comment (qc00) [reply cpp/arcticdb/entity/performance_tracing.cpp:16]: As long as you have considered that alternative and there's written record of the deliberation (here), I am fine with either solution.

Comment (jjerphan) [reply cpp/arcticdb/entity/performance_tracing.cpp:16]: I think that I prefer this alternative. I have also removed the useless noise with c50cde9. 

In the long term, I think it would be valuable to move custom tools' setup (like Remotery) entirely to dedicated branches rebased on `master` to ease the CMake setup. What do you think?

For now, I let you approve and merge if this looks good to you to make the decision clear and consensual.

**Review by qc00 — APPROVED** [review]

**Comment by jjerphan** [comment]
As discussed with @qc00 and @mehertz, merging with one approval is common practice for now (unless decided otherwise in the future). I will merge this PR and other approved ones.

**Pull request merged**
